src/v/datalake/values_protobuf.cc

```diff
@@ -0,0 +1,333 @@
+          [&stack, &field_descriptor](auto& list) {
+              return convert_repeated(std::move(list), field_descriptor, stack);
+          });
+    }
+
+    return single_field_to_value(std::move(field), field_descriptor, stack);
+}
+
+ss::future<value_outcome> single_field_to_value(
+  std::optional<parsed::message::field> field,
+  const pb::FieldDescriptor& field_descriptor,
+  proto_descriptors_stack& stack) {
+    switch (field_descriptor.type()) {
+    case pb::FieldDescriptor::TYPE_DOUBLE:
+        co_return convert<double, iceberg::double_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_double);
+    case pb::FieldDescriptor::TYPE_FLOAT:
+        co_return convert<float, iceberg::float_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_float);
+    // casting uint32 to long value to prevent overflow
+    case pb::FieldDescriptor::TYPE_UINT32:
+    case pb::FieldDescriptor::TYPE_FIXED32:
+    case pb::FieldDescriptor::TYPE_SFIXED64:
+    case pb::FieldDescriptor::TYPE_INT64:
+    case pb::FieldDescriptor::TYPE_SINT64:
+        co_return convert<int64_t, iceberg::long_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_int64);
+    // unsigned 64 bit integers are not supported
+    case pb::FieldDescriptor::TYPE_UINT64:
+    case pb::FieldDescriptor::TYPE_FIXED64:
+        co_return type_conversion_error(field_descriptor);
+    case pb::FieldDescriptor::TYPE_INT32:
+    case pb::FieldDescriptor::TYPE_SFIXED32:
+    case pb::FieldDescriptor::TYPE_SINT32:
+        co_return convert<int32_t, iceberg::int_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_int32);
+    case pb::FieldDescriptor::TYPE_ENUM:
+        if (field.has_value()) {
+            co_return iceberg::int_value(
+              std::get<int32_t>(std::move(field.value())));
+        }
+        co_return iceberg::int_value(
+          field_descriptor.default_value_enum()->number());
+    case pb::FieldDescriptor::TYPE_BOOL:
+        co_return convert<bool, iceberg::boolean_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_bool);
+    case pb::FieldDescriptor::TYPE_STRING:
+        co_return convert<iobuf, iceberg::string_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_string);
+    case pb::FieldDescriptor::TYPE_GROUP:
+        co_return type_conversion_error(field_descriptor);
+    case pb::FieldDescriptor::TYPE_MESSAGE:
+        if (!field.has_value()) {
+            co_return co_await message_to_value(
+              std::make_unique<parsed::message>(),
+              *field_descriptor.message_type(),
+              stack);
+        }
```

Comment:
Is this the right semantics for Iceberg? For protobuf there is a semantic difference between missing and default values for messages. Only primitives really have "defaults". I'm trying to figure out if we will ever have `null` values because AFAIK with this approach we will not.

Reply:
This is what i was wondering as well how to treat this. In a `parsed::message` fields that are not set are simply not there, a question is how to treat them, in `struct_value` we have a vector of optionals, if that is what it should be we can simply return an empty optional in this case.

Reply:
Yeah I believe that we want an `std::nullopt` for this case. Also, we should consider handling explicitly "optional" fields in proto2 and the new field presence stuff in proto3. IMO proto2 optional values that are not set should probably be marked as null - and same if a field is marked as optional in proto3 with field presence enabled.

sorry the parser kind of punts this onto the user of the interface to get right, maybe we can consider having the parser handle the `null` vs `default` case.

References:

* https://protobuf.dev/programming-guides/proto2/#optional
* https://protobuf.dev/programming-guides/proto3/#default
* https://protobuf.dev/programming-guides/field_presence/
* https://protobuf.dev/programming-guides/field_presence/#cheat